### ISSUE TITLE
trouble running nteract via CLI

### ISSUE THREAD

**Issue by AlJohri** [nteract/archived-desktop-app#1576]
Hi, I'm having some trouble running nteract via the CLI.

I've tried running nteract from the cli like so:

```
nteract
```

it starts the process in the background and shows the loading screen but gets stuck there for quite a long time. I've never seen it finish.

When I run

```
nteract --verbose
```

it starts the process in the foreground and seems to work properly in a timely manner.

I tried debugging further and see that when I run it in the foreground with verbose, this is the log:

```
$ nteract --verbose
[12:55:11:0037] [info] args { _: [],
  version: false,
  v: false,
  help: false,
  h: false,
  verbose: true,
  kernel: 'python3',
  k: 'python3',
  '$0': '/Applications/nteract.app/Contents/MacOS/nteract --verbose' }
[12:55:14:0483] [info] launching an empty notebook by default
```

When I try to run it in the background without verbose, I checked the nohup.out file and this is the log:

```
$ nteract
$ cat /tmp/nteract-nohup.out
[12:54:37:0680] [info] args { _: [],
  version: false,
  v: false,
  help: false,
  h: false,
  kernel: 'python3',
  k: 'python3',
  '$0': '/Applications/nteract.app/Contents/MacOS/nteract',
  verbose: undefined }
[12:54:52:0172] [info] launching an empty notebook by default
```

I can't see any difference really. Looking through the Mac `Console` app I also don't see anything of interest.

Any ideas on how to debug further?

Thanks!

**Comment by lgeiger** [comment]
It looks like for some reason the `nohup` command doesn't work on your machine. Which shell are you using?

If you feel adventurous you could install the development version of nteract and try out [this branch](https://github.com/lgeiger/nteract/tree/cli). It uses a different method for spawning and detaching nteract from the terminal.

**Comment by AlJohri** [comment]
@lgeiger I've used `nohup` a bit on my laptop and it should work. Doesn't the fact that I see `/tmp/nteract-nohup.out` with updated timestamps mean that nohup is working properly?

I'll try out the development version when I get a chance.

**Comment by AlJohri** [comment]
@lgeiger great that worked for me!

I ran:

```
npm install
npm dist
```

And then opened the dmg and dragged `nteract` to the `/Applications` folder.

I was able to start by:
- typing `nteract` in the CLI
- typing `nteract Untitled.ipynb` in the CLI
- clicking `Untitled.ipynb` in Finder
- clicking on the app itself in `/Applications` via Finder/Dock

Notes:
- using the CLI it always creates a separate dock icon each time you type `nteract` with or without a filename; clicking on the filename in Finder or App itself this doesn't occur
- regardless of whether starting the app from the CLI or GUI, I am always asked: 'Do you want the application "nteract Helper.app" to accept incoming network connections?'
  - this may be because its a local unsigned installation though
![screen shot 2017-03-23 at 3 50 37 pm](https://cloud.githubusercontent.com/assets/2790092/24267506/e06e58b6-0fe0-11e7-8d7e-fd1ece08d806.png)

**Comment by rgbkrk** [comment]
> Do you want the application "nteract Helper.app" to accept incoming network connections?

That's the mac firewall (you won't see this if you turn it off). After approving the first time for a signed app, it won't show up again. Since, as you said, every opening of the unsigned app creates a separate dock icon, this annoying dialogue comes up.

**Comment by AlJohri** [comment]
Cool. Is there anything else needs to be done before the @lgeiger `cli` branch can be rebased and merged in? I can make a PR if the rebase doesn't turn out to be to complicated.

**Comment by lgeiger** [comment]
@AlJohri Excellent! Thanks a lot for trying!

Please go ahead and make a PR. It shouldn't need a rebase, I rebased it onto master before I sent you the link 😉 

> using the CLI it always creates a separate dock icon each time you type nteract with or without a filename; clicking on the filename in Finder or App itself this doesn't occur

You're right, that's annoying but at the moment expected behavior. We should definitely improve this.

**Comment by AlJohri** [comment]
closing in favor of #1581

**Comment by AlJohri** [comment]
Just came across the pycharm run script which has some logic for checking if the app is already running. Throwing a copy here in case its helpful: https://gist.github.com/AlJohri/67565210406871d4ac25c5bf5834d239

**Comment by rgbkrk** [comment]
Would be cool to see that integrated @AlJohri! We just shipped 0.2.0 and I'm putting together release notes now.

**Comment by AlJohri** [comment]
@rgbkrk one benefit of the way the process starts now is that it inherits the environment variables of the current shell. if the cli integrates with an existing nteract instance, what do you think would be best practice for env vars?

my main use case is database credentials - I use notebooks pretty often to store mongo/sql queries

**Comment by rgbkrk** [comment]
> my main use case is database credentials - I use notebooks pretty often to store mongo/sql queries

That makes sense.

> if the cli integrates with an existing nteract instance, what do you think would be best practice for env vars?

I can't think of a good way really. Stewing on it a bit, one way to do it is to set environment variables in your kernelspec file. 

------------------------

There are some drawbacks of multiple instances of the app, primarily taking up more memory than necessary (which is pretty big for these electron apps) as well as making it really confusing when viewing it in the switcher (how do you figure out which one has the notebook you're working on).

It's pretty clear we've got some tradeoffs here. ¯\\\_(ツ)\_/¯

**Comment by AlJohri** [comment]
I'd be interested to explore a little more how other GUI apps do it.

- experimenting with pycharm, starting the app via the command line using `charm .` will take the environment variables of the current shell. trying to start pycharm again via the command line in either the same terminal session or another has no affect on the environment variables - they are frozen after first launch. this makes sense

- pycharm also has an interface to set environment variables at a project level but its kind of a pain to use: 
![screen shot 2017-09-23 at 11 40 45 pm](https://user-images.githubusercontent.com/2790092/30779224-a80d55ee-a0b8-11e7-8e7c-1a30ced7e400.png)

- pycharm also has a EnvFile plugin which is what I currently use: 
![screen shot 2017-09-23 at 11 41 31 pm](https://user-images.githubusercontent.com/2790092/30779225-c135533c-a0b8-11e7-82f1-31290b111f22.png) it works fine but I'm not a big fan of env files - I prefer using [envdir](https://github.com/jezdez/envdir) so its inconvenient to have to manage both. but perhaps that's more of a personal opinion, I suppose :)

- there is always `launchctl setenv VAR VALUE` which I think mostly works but that is seriously a nightmare and a half

I'll ponder on this some more, but I think telling someone to quit nteract and start it again from the command line to get the env vars is not the worst thing in the world for now since that should work OOB

**Comment by AlJohri** [comment]
Seems like R uses `Renviron` files but even Rstudio has some issues with this. Can only set like global environment variables not project / workspace specific.
